packages/xl-ai/src/streamTool/streamTool.ts

```diff
@@ -0,0 +1,96 @@
+import { DeepPartial } from "ai";
+import type { JSONSchema7 } from "json-schema";
+
+export type InvalidOrOk<T> =
+  | {
+      result: "invalid";
+      reason: string;
+    }
+  | { result: "ok"; value: T };
+
+/**
+ * A StreamTool is a function that can be called by the LLM.
+ * It's similar to a Tool in the Vercel AI SDK, but:
+ *
+ * - a collection of StreamTools can be wrapped in a single LLM Tool to issue multiple operations (tool calls) at once.
+ * - StreamTools can be used in a streaming manner.
+ */
+export type StreamTool<T extends { type: string }> = {
+  /**
+   * The name of the tool
+   */
+  name: T["type"];
+  /**
+   * An optional description of the tool that can influence when the tool is picked.
+   */
+  description?: string;
+  /**
+   * The schema of the input that the tool expects. The language model will use this to generate the input.
+   */
+  parameters: JSONSchema7;
+  /**
+   * Validates the input of the tool call
+   *
+   * @param operation - The operation to validate.
+   * This can be a partial object as the LLM may not have completed the operation yet.
+   * The object is not guaranteed to be of type DeepPartial<T> as the LLM might not conform to the schema correctly - so this needs to be validated.
+   */
+  validate: (
+    operation: DeepPartial<T>, // TODO: maybe `unknown` is better?
+  ) => InvalidOrOk<T>;
+
+  /**
+   * Executes tool calls.
+   *
+   * Note that this operates on the full stream of operations (including other StreamTools calls).
+   * This way, we can
+   *
+   * @returns the stream of operations that have not been processed (and should be passed on to execute handlers of other StreamTools)
+   */
+  execute: (
+    operationsStream: AsyncIterable<{
+      operation: StreamToolCall<StreamTool<any>[]>;
+      isUpdateToPreviousOperation: boolean;
+      isPossiblyPartial: boolean;
+    }>,
+  ) => AsyncIterable<{
+    operation: StreamToolCall<StreamTool<any>[]>;
+    isUpdateToPreviousOperation: boolean;
+    isPossiblyPartial: boolean;
```

Comment:
I don't know about these names, but I can't think of anything better.

isUpdateToPreviousOperation -> amend?
isPossiblyPartial -> isComplete?

IDK

Reply:
TODO: comment

Reply:
documented here https://github.com/TypeCellOS/BlockNote/blob/69d06f2eb392387484aa9aec99b8e0ebcaa67f54/packages/xl-ai/src/streamTool/callLLMWithStreamTools.ts#L54